walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
+
+    // after all events are processed, for each minute x, meetings[x] will represent the number of
+    // meetings that start at minute x, minus the number of meetings that end right before minute x
+
+    // precompute the prefix sum on the meetings array s.t meetings[x] will represent the number of
+    // meetings that happen at minute x:
+    // (before computing the sum):  [ .... +1 ........ -1 .... ]
+    // (after computing the sum) :  [..... +1 +1 ... +1 0 .... ]
+    precomputePrefixSum(meetings);
+
+    return findAvailableTimeRanges(meetings, (int)request.getDuration());
+  }
+
+  /** In meetings array, add 1 to the start time of the meeting and substract 1 from the end time
+    * Only the endpoints are changed s.t. after all events are processed and the prefix sum is
+    * computed, the number of meetings increases in the array starting from start time and ending
+    * right before the end time
+   */
+  private void updateNumberOfMeetings(ArrayList<Integer> meetings, TimeRange when) {
+    // mark that a new meeting starts at when.start()
+    meetings.set(when.start(), meetings.get(when.start()) + 1); 
+    // mark that a new meeting ends right before when.end() 
+    if (when.end() <= TimeRange.END_OF_DAY) {
+      // only update the end time of events that end before the day ends
+      meetings.set(when.end(), meetings.get(when.end()) - 1);
+    }
+  }
+
+  private void precomputePrefixSum(ArrayList<Integer> meetings) {
+      // compute the sum starting from the second element because the prefix sum for the first 
+      // element is the first element 
+      for (int i = TimeRange.START_OF_DAY + 1; i <= TimeRange.END_OF_DAY; ++ i) {
+        meetings.set(i, meetings.get(i - 1) + meetings.get(i));
+      }
+  }
+
+  /** Given meetings = an array where each element represents the number of meeting that occur at that
+   * minute, and a duration, returns a Collection of time ranges in which the meeting lasting duration
+   * minutes can happen 
+  */
+  private Collection<TimeRange> findAvailableTimeRanges(ArrayList<Integer> meetings, int duration) {
+    ArrayList<TimeRange> availableTimeRange = new ArrayList<TimeRange>();
+
+    int lastUnavailableTime = TimeRange.START_OF_DAY - 1;
+    
+    for (int endingTime = TimeRange.START_OF_DAY; endingTime <= TimeRange.END_OF_DAY; ++ endingTime) {
+      if (meetings.get(endingTime) != 0) {
+        // at least one meeting is scheduled during minute endingTime
+        lastUnavailableTime = endingTime;
+      }
+      if (endingTime == TimeRange.END_OF_DAY || meetings.get(endingTime + 1) != 0) {
+        // there are no meetings during (lastUnavailableTime, endingTime]
+        if (endingTime - lastUnavailableTime >= duration) {
+          // add [lastUnavailableTime + 1, endingTime] as an available time range for the meeting
+          TimeRange timeRange = TimeRange.fromStartEnd(lastUnavailableTime + 1, endingTime, true);
```

Comment:
Regarding this true at the end, shamelessly copy pasting some readability best practice:

Document ambiguous literal arguments at the call site
In certain contexts literal arguments - such as 0, "", true and false, or null - can make it difficult for readers to know what a method will do. We generally discourage defining methods that take boolean parameters or otherwise expect users to pass in ambiguous literals. However, when you must call such a method you're encouraged to use the parameter name as an inline comment at the call site, so that readers don't need to look at the method declaration to understand the parameter's purpose.

Such a comment should be formatted in this way:

booleanMethod(/* enableFoo= */ true);